lib/puma/server.rb

```diff
@@ -278,27 +281,15 @@ def handle_servers
               if sock == check
                 break if handle_check
               else
-                begin
-                  if io = sock.accept_nonblock
-                    client = Client.new io, @binder.env(sock)
-                    if remote_addr_value
-                      client.peerip = remote_addr_value
-                    elsif remote_addr_header
-                      client.remote_addr_header = remote_addr_header
-                    end
-
-                    pool << client
-                    pool.wait_until_not_full
-                  end
-                rescue SystemCallError
-                  # nothing
-                rescue Errno::ECONNABORTED
-                  # client closed the socket even before accept
-                  begin
-                    io.close
-                  rescue
-                    Thread.current.purge_interrupt_queue if Thread.current.respond_to? :purge_interrupt_queue
+                pool.wait_until_not_full
```

Comment:
Moving the call to `wait_until_not_full` before `accept_socket` prevents connections from being accepted until out_of_band hooks are completed.

(I'm not 100% sure this change won't have any other side-effects, but the test suite seems to think it's okay...)

An alternative would be to just add an extra `pool.with_mutex {}` line above `accept_socket`.